spv/sync.go

```diff
@@ -607,9 +616,14 @@ func (s *Syncer) connectToCandidates(ctx context.Context) error {
 
 		wg.Add(1)
 		go func() {
+			defer wg.Done()
 			raddr := na.String()
-			s.connectAndRunPeer(ctx, raddr, false)
-			wg.Done()
+			backoff := s.connectAndRunPeer(ctx, raddr, false)
+			if backoff {
+				// Delay next connect attempt to save resources and not overwhelm peers,
+				// it's unlikely to succeed anyway if we do retry immediately.
+				time.Sleep(5 * time.Second)
+			}
```

Comment:
This is blocking. If you want to sleep you need to use a pattern like above:
```
		        select {
			case <-ctx.Done():
				return
			case <-time.After(5 * time.Second):
			}
```

I think a better way to solve this is to detect that we are not connected and pause this loop until we think we are connected. Unsure how to go about that though. Maybe a combination of sniffing out the error `network is unreachable` and sleeps is the easiest. Unless there's a better way to know if connected to the network.

Reply:
Yeah, maybe what you are doing is best. If you could try to sniff out that particular error and change the sleeps I think this is ok. ofc jrick may have better ideas.

Reply:
> If you could try to sniff out that particular error and change the sleeps I think this is ok

Yeah, essentially it would be best to classify all the errors that can happen in `connectAndRunPeer` in two groups (backoff == true/false).

Reply:
It seems to me like this should mirror the [backoff functionality for permanent peers in dcrd](https://github.com/decred/dcrd/blob/6fd343ab08d489034e4a69742dbe0c3e2b700620/connmgr/connmanager.go#L255-L268), which is actually a backoff versus a static retry every 5 seconds.

The code structure is obviously a bit different here, but the key point is that it keeps a retry count and increases the backup by the configured number of seconds (5 seconds here) times the retry count and is clamped to a maximum so it doesn't grow indefinitely (that max 5 minutes in dcrd).

Reply:
Another thing that dcrd does is keep a global failed attempts counter for non-persistent peers that is reset to 0 any time there is a successful connection.  Whenever that global failed attempt counter reaches or exceeds a maximum number of failed attempts (dcrd uses 25), it imposes a static delay with the retry duration (5 seconds) on all non-persistent connection attempts too.

That ensures when the network goes down, it doesn't go nuts trying to connect thousands of times per second due to the connection attempts immediately failing.

Reply:
> In other words, it'll sleep 5 seconds, then 10 seconds, then 15 seconds, and continue increasing by 5 seconds all the way to 5 minutes.

Yep, essentially this protects us from overwhelming our peer(s), so that they don't ban us (or just have too much connections/requests which slows them) ?

And yes, if timeous keep increasing, at some point we want to reset them or something to not starve ourselves of connections for too long.

I certainly can implement that, but probably better get some feedback on this from @jrick first to see if I implemented the `backoff` return value (from `connectAndRunPeer`) correctly.